## Question 06

> What is the most common speaker occupation in the dataset?

In [1]:
import pandas as pd
import json

CSV = "https://raw.githubusercontent.com/cpethe/TED_Talks/master/ted_main.csv"
df = pd.read_csv(CSV)

df['speaker_occupation'].value_counts().idxmax()

'Writer'

## Question 07

> Drop a column from the dataframe that is uninformative (information already contained in other columns), so that the dataframe no longer contains that column.



In [2]:
if 'name' in df.columns:
  df.drop('name', axis=1, inplace=True)

## Question 08

> Get the rows corresponding to talks about climate change.

In [3]:
talks_abt_climate_change = df[df["tags"].apply(lambda x: "climate change" in x)]
talks_abt_climate_change

comments                                        description  duration  \
1          265  With the same humor and humanity he exuded in ...       977   
25         184  Legendary scientist David Deutsch puts theoret...      1140   
38          57  Arctic explorer Ben Saunders recounts his harr...      1083   
51         499  Given $50 billion to spend, which would you so...      1001   
54         203  Speaking as both an astronomer and "a concerne...      1046   
...        ...                                                ...       ...   
2478        31  Anab Jain brings the future to life, creating ...       881   
2486        12  Rivers are one of nature's most powerful force...       668   
2488        26  Climate change is real, case closed. But there...       787   
2497        17  Corals in the Pacific Ocean have been dying at...       434   
2534         2  What the astronauts felt when they saw Earth f...       725   

               event   film_date  languages      main_speaker  num_speaker  \
1            TED2006  1140825600         43           Al Gore            1   
25    TEDGlobal 2005  1121299200         29     David Deutsch            1   
38           TED2005  1109203200         26      Ben Saunders            1   
51           TED2005  1107302400         32     Bjorn Lomborg            1   
54    TEDGlobal 2005  1121299200         29       Martin Rees            1   
...              ...         ...        ...               ...          ...   
2478         TED2017  1492992000         10         Anab Jain            1   
2486         TEDxPSU  1393718400         12         Liz Hajek            1   
2488         TED2017  1492992000         10       Kate Marvel            1   
2497         TED2017  1492992000         12  Kristen Marhaver            1   
2534       TEDxSkoll  1491523200          1    Benjamin Grant            1   

      published_date                                            ratings  \
1         1151367060  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
25        1158019860  [{'id': 9, 'name': 'Ingenious', 'count': 269},...   
38        1161735060  [{'id': 7, 'name': 'Funny', 'count': 80}, {'id...   
51        1167696660  [{'id': 3, 'name': 'Courageous', 'count': 283}...   
54        1168992660  [{'id': 1, 'name': 'Beautiful', 'count': 214},...   
...              ...                                                ...   
2478      1497884701  [{'id': 1, 'name': 'Beautiful', 'count': 47}, ...   
2486      1499957123  [{'id': 10, 'name': 'Inspiring', 'count': 11},...   
2488      1500303942  [{'id': 24, 'name': 'Persuasive', 'count': 20}...   
2497      1501253483  [{'id': 23, 'name': 'Jaw-dropping', 'count': 1...   
2534      1504814438  [{'id': 10, 'name': 'Inspiring', 'count': 46},...   

                                          related_talks    speaker_occupation  \
1     [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...      Climate advocate   
25    [{'id': 2237, 'hero': 'https://pe.tedcdn.com/i...     Quantum physicist   
38    [{'id': 2292, 'hero': 'https://pe.tedcdn.com/i...       Arctic explorer   
51    [{'id': 248, 'hero': 'https://pe.tedcdn.com/im...    Global prioritizer   
54    [{'id': 167, 'hero': 'https://pe.tedcdn.com/im...        Astrophysicist   
...                                                 ...                   ...   
2478  [{'id': 2858, 'hero': 'https://pe.tedcdn.com/i...    Futurist, designer   
2486  [{'id': 2424, 'hero': 'https://pe.tedcdn.com/i...          Geoscientist   
2488  [{'id': 1763, 'hero': 'https://pe.tedcdn.com/i...     Climate scientist   
2497  [{'id': 2385, 'hero': 'https://pe.tedcdn.com/i...  Coral reef biologist   
2534  [{'id': 2511, 'hero': 'https://pe.tedcdn.com/i...                Author   

                                                   tags  \
1     ['alternative energy', 'cars', 'climate change...   
25    ['climate change', 'cosmos', 'culture', 'envir...   
38    ['climate change', 'culture', 'exploration', '...   
51    ['AI

## Question 09

> Get the rows corresponding to 10 most lengthy talks with at least 10 million views or at least 3000 comments

In [4]:
df[(df["views"] >= 10_000_000) | (df["comments"] >= 3000)].sort_values(by="duration", ascending=False).head(10)

comments                                        description  duration  \
96        6404  Richard Dawkins urges all atheists to openly s...      1750   
644       3356  Questions of good and evil, right and wrong ar...      1386   
1940      1355  "Public shaming as a blood sport has to stop,"...      1346   
5          672  Tony Robbins discusses the "invisible forces" ...      1305   
29         970  Dan Gilbert, author of "Stumbling on Happiness...      1276   
1346      2290  Body language affects how others see us, but i...      1262   
837       1927  Brené Brown studies human connection -- our ab...      1219   
596        296  In this highly personal talk from TEDMED, magi...      1219   
4          593  You've never seen data presented like this. Wi...      1190   
262        669  First, Keith Barry shows us how our brains can...      1189   

               event   film_date  languages     main_speaker  num_speaker  \
96           TED2002  1012608000         42  Richard Dawkins            1   
644          TED2010  1265846400         39       Sam Harris            1   
1940         TED2015  1426723200         41  Monica Lewinsky            1   
5            TED2006  1138838400         36     Tony Robbins            1   
29           TED2004  1075680000         43      Dan Gilbert            1   
1346  TEDGlobal 2012  1340668800         51        Amy Cuddy            1   
837      TEDxHouston  1275782400         52      Brené Brown            1   
596      TEDMED 2009  1256601600         34     David Blaine            1   
4            TED2006  1140566400         48     Hans Rosling            1   
262          TED2004  1075680000         28      Keith Barry            1   

      published_date                                            ratings  \
96        1176689220  [{'id': 3, 'name': 'Courageous', 'count': 3236...   
644       1269249180  [{'id': 8, 'name': 'Informative', 'count': 923...   
1940      1426894031  [{'id': 3, 'name': 'Courageous', 'count': 8668...   
5         1151440680  [{'id': 7, 'name': 'Funny', 'count': 1102}, {'...   
29        1159229460  [{'id': 7, 'name': 'Funny', 'count': 1728}, {'...   
1346      1349103608  [{'id': 23, 'name': 'Jaw-dropping', 'count': 3...   
837       1293115500  [{'id': 10, 'name': 'Inspiring', 'count': 2144...   
596       1263889320  [{'id': 22, 'name': 'Fascinating', 'count': 91...   
4         1151440680  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   
262       1216366800  [{'id': 2, 'name': 'Confusing', 'count': 273},...   

                                          related_talks  \
96    [{'id': 86, 'hero': 'https://pe.tedcdn.com/ima...   
644   [{'id': 666, 'hero': 'https://pe.tedcdn.com/im...   
1940  [{'id': 2073, 'hero': 'https://pe.tedcdn.com/i...   
5     [{'id': 229, 'hero': 'https://pe.tedcdn.com/im...   
29    [{'id': 944, 'hero': 'https://pe.tedcdn.com/im...   
1346  [{'id': 605, 'hero': 'https://pe.tedcdn.com/im...   
837   [{'id': 1391, 'hero': 'https://pe.tedcdn.com/i...   
596   [{'id': 310, 'hero': 'https://pe.tedcdn.com/im...   
4     [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   
262   [{'id': 1821, 'hero': 'https://pe.tedcdn.com/i...   

                               speaker_occupation  \
96                         Evolutionary biologist   
644                   Neuroscientist, philosopher   
1940                              Social activist   
5     Life coach; expert in leadership psychology   
29                 Psychologist; happiness expert   
1346                          Social psychologist   
837                      Vulnerability researcher   
596                 Illusionist, endurance artist   
4            Global health expert; data visionary   
262                                      Magician   

                                                   tags  \
96    ['God', 'atheism', 'culture', 'religion', 'sci...   
644   ['culture', 'evolutionary psychology', 'global...   
1940  ['communication', 'media', 'social media', 's

## Question 10

> Which talk is the most viewed as compared to its *related talks* (the one with the maximum difference between its views and the *view count* of any of its *related talks*)?

In [5]:
import ast

def compute_related_diff(row):
    related_views = []
    for related in ast.literal_eval(row["related_talks"]):
        related_views.append(df[df["title"] == related["title"]]["views"].values[0])
    return max([row["views"] - x for x in related_views]) if related_views else 0

df["related_diff"] = df.apply(compute_related_diff, axis=1)

# Sort DataFrame by the computed difference in descending order
df = df.sort_values(by="related_diff", ascending=False)

df.iloc[0][:-1]

comments                                                           4553
description           Sir Ken Robinson makes an entertaining and pro...
duration                                                           1164
event                                                           TED2006
film_date                                                    1140825600
languages                                                            60
main_speaker                                               Ken Robinson
num_speaker                                                           1
published_date                                               1151367060
ratings               [{'id': 7, 'name': 'Funny', 'count': 19645}, {...
related_talks         [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...
speaker_occupation                                      Author/educator
tags                  ['children', 'creativity', 'culture', 'dance',...
title                                       Do schools kill creativity?
url                   https://www.ted.com/talks/ken_robinson_says_sc...
views                                                          47227110
Name: 0, dtype: object

## Question 11

> Which pair of features (columns) are the most correlated? Comment on whether the correlation implies causation in this case with a few sentences.

In [6]:
max_correlated, max_correlation = max(
    [
        ((col1, col2), df[col1].corr(df[col2]))
        for col1 in df.select_dtypes(exclude="object").columns
        for col2 in df.select_dtypes(exclude="object").columns
        if col1 != col2
    ],
    key=lambda x: x[1],
)
max_correlated, max_correlation

(('related_diff', 'views'), 0.9784920041585558)

The correlation in this case does not imply causation. Both dates are bound by a shared timeline rather than a causal relationship. The film date doesnt *cause* the publication; rather, both variables are dependent on decisions made within a larger production process. The dates being sequential makes them correlated but this correlation merely relfects that they are part of a shared timeline -- not that one event is causing the other.

## Question 12

> Research a way to find out the significance (in terms of p-value) of the correlation of a pair of features. Try it on the pair of columns *duration* and *comments*. What is the correlation coefficient and its p-value? Comment on what this finding implies with a few sentences.

In [7]:
from scipy.stats import pearsonr

correlation, p_value = pearsonr(df["duration"], df["comments"])

print(f"Correlation {correlation:.4f}\np-value: {p_value:.13f}")

Correlation 0.1407
p-value: 0.0000000000010


The p-value here represents the probability under the null hypothesis (the two features are uncorrelated) of obtaining a correlation as or more extreme than the one computed from the datasets. Since the probability of obtaining a correlation of 0.1407 is ~0.0000000000010, it's safe to reject the null hypothesis and conclude that the correlation between duration and comments is statistically significant. I.e., the correlation is not due to random chance and exists earnestly in the population of TED talks.